### ISSUE TITLE
[BUG] Can't run on Raspberry Pi OS (Bookworm)

### ISSUE THREAD

**Issue by probonopd** [waydroid/waydroid#1171]
### Describe the bug

Can't run on Raspberry Pi OS (Bookworm) on Raspberry Pi 5.

```
pip install pyclip --break-system-packages
sudo waydroid upgrade --offline

sudo rm /var/lib/waydroid/waydroid.log

sudo systemctl restart waydroid-container.service

waydroid show-full-ui

[13:45:20] Starting waydroid session
[gbinder] Service manager /dev/binder has appeared
[gbinder] WARNING: Service manager /dev/binder has died
```

### Waydroid version

1.4.2

### Device

Linux Desktop

### Operating System

Raspberry Pi OS based on Debian GNU/Linux 12 (bookworm)

### Kernel version

6.1.0-rpi6-rpi-v8

### Desktop Environment

Raspberry Pi OS version of lxde

### GPU

Raspberry Pi 5 GPU

### Logs

/var/lib/waydroid/waydroid.cfg

```
[waydroid]
arch = arm64
vendor_type = MAINLINE
system_datetime = 1700244220
vendor_datetime = 1700261635
suspend_action = freeze
mount_overlays = True
images_path = /var/lib/waydroid/images
system_ota = https://ota.waydro.id/system/lineage/waydroid_arm64/VANILLA.json
vendor_ota = https://ota.waydro.id/vendor/waydroid_arm64/MAINLINE.json
binder = binder
vndbinder = vndbinder
hwbinder = hwbinder
binder_protocol = aidl3
service_manager_protocol = aidl3

[properties]
````

/var/lib/waydroid/waydroid.prop

```
sys.use_memfd=true
debug.stagefright.ccodec=0
ro.hardware.gralloc=gbm
ro.hardware.egl=mesa
ro.hardware.camera=v4l2
ro.opengles.version=196609
waydroid.system_ota=https://ota.waydro.id/system/lineage/waydroid_arm64/VANILLA.json
waydroid.vendor_ota=https://ota.waydro.id/vendor/waydroid_arm64/MAINLINE.json
waydroid.tools_version=1.4.2
ro.vndk.lite=true
ro.product.waydroid.model=Raspberry Pi 5 Model B Rev 1.0
waydroid.host.user=user
waydroid.host.uid=1000
waydroid.host.gid=1000
waydroid.host_data_path=/home/user/.local/share/waydroid/data
waydroid.xdg_runtime_dir=/run/user/1000
waydroid.pulse_runtime_path=/run/user/1000/pulse
waydroid.wayland_display=wayland-1
waydroid.background_start=false
waydroid.stub_sensors_hal=1
```

/var/lib/waydroid/waydroid_base.prop

```
sys.use_memfd=true
debug.stagefright.ccodec=0
ro.hardware.gralloc=gbm
ro.hardware.egl=mesa
ro.hardware.camera=v4l2
ro.opengles.version=196609
waydroid.system_ota=https://ota.waydro.id/system/lineage/waydroid_arm64/VANILLA.json
waydroid.vendor_ota=https://ota.waydro.id/vendor/waydroid_arm64/MAINLINE.json
waydroid.tools_version=1.4.2
ro.vndk.lite=true
ro.product.waydroid.model=Raspberry Pi 5 Model B Rev 1.0
```

/var/lib/waydroid/waydroid.log

```
(005909) [13:45:19] % lxc-info -P /var/lib/waydroid/lxc -n waydroid -sH
STOPPED
(005909) [13:45:19] % modprobe -q ashmem_linux
(005909) [13:45:19] % chmod 666 -R /dev/binder
(005909) [13:45:19] % chmod 666 -R /dev/vndbinder
(005909) [13:45:19] % chmod 666 -R /dev/hwbinder
(005913) [13:45:20] Starting waydroid session
(005909) [13:45:20] % /usr/lib/waydroid/data/scripts/waydroid-net.sh start
vnic is waydroid0
(005909) [13:45:20] % systemctl is-active -q nfcd
(005909) [13:45:20] % chmod 777 -R /dev/dri
(005909) [13:45:20] % chmod 777 -R /dev/fb0
(005909) [13:45:20] % chmod 777 -R /dev/video19
(005909) [13:45:20] % chmod 777 -R /dev/video37
(005909) [13:45:20] % chmod 777 -R /dev/video36
(005909) [13:45:20] % chmod 777 -R /dev/video35
(005909) [13:45:20] % chmod 777 -R /dev/video34
(005909) [13:45:20] % chmod 777 -R /dev/video33
(005909) [13:45:20] % chmod 777 -R /dev/video32
(005909) [13:45:20] % chmod 777 -R /dev/video31
(005909) [13:45:20] % chmod 777 -R /dev/video30
(005909) [13:45:20] % chmod 777 -R /dev/video29
(005909) [13:45:20] % chmod 777 -R /dev/video28
(005909) [13:45:20] % chmod 777 -R /dev/video27
(005909) [13:45:20] % chmod 777 -R /dev/video26
(005909) [13:45:20] % chmod 777 -R /dev/video25
(005909) [13:45:20] % chmod 777 -R /dev/video24
(005909) [13:45:20] % chmod 777 -R /dev/video23
(005909) [13:45:20] % chmod 777 -R /dev/video22
(005909) [13:45:20] % chmod 777 -R /dev/video21
(005909) [13:45:20] % chmod 777 -R /dev/video20
(005909) [13:45:20] % mv /var/lib/waydroid/config_session /var/lib/waydroid/lxc/waydroid
(005909) [13:45:20] % mount -o ro /var/lib/waydroid/images/system.img /var/lib/waydroid/rootfs
(005909) [13:45:20] % mount -t overlay -o ro,lowerdir=/var/lib/waydroid/overlay:/var/lib/waydroid/rootfs,upperdir=/var/lib/waydroid/overlay_rw/system,workdir=/var/lib/waydroid/overlay_work/system,xino=off overlay /var/lib/waydroid/rootfs
(005909) [13:45:20] % mount -o ro /var/lib/waydroid/images/vendor.img /var/lib/waydroid/rootfs/vendor
(005909) [13:45:20] % mount -t overlay -o ro,lowerdir=/var/lib/waydroid/overlay/vendor:/var/lib/waydroid/rootfs/vendor,upperdir=/var/lib/waydroid/overlay_rw/vendor,workdir=/var/lib/waydroid/overlay_work/vendor,xino=off overlay /var/lib/waydroid/rootfs/vendor
(005909) [13:45:20] % mount -o bind /var/lib/waydroid/waydroid.prop /var/lib/waydroid/rootfs/vendor/waydroid.prop
(005909) [13:45:20] Save config: /var/lib/waydroid/waydroid.cfg
(005909) [13:45:20] % lxc-start -P /var/lib/waydroid/lxc -F -n waydroid -- /init
(005909) [13:45:20] New background process: pid=6006, output=background
(005909) [13:45:20] % lxc-info -P /var/lib/waydroid/lxc -n waydroid -sH
RUNNING
lxc-start: waydroid: ../src/lxc/conf.c: run_buffer: 322 Script exited with status 126
lxc-start: waydroid: ../src/lxc/start.c: lxc_end: 985 Failed to run lxc.hook.post-stop for container "waydroid"
(005909) [13:48:29] % lxc-info -P /var/lib/waydroid/lxc -n waydroid -sH
STOPPED
(005909) [13:48:29] % /usr/lib/waydroid/data/scripts/waydroid-net.sh stop
vnic is waydroid0
(005909) [13:48:29] % systemctl is-enabled -q nfcd
Failed to get unit file state for nfcd.service: No such file or directory
(005909) [13:48:29] % umount /var/lib/waydroid/rootfs/vendor/waydroid.prop
(005909) [13:48:29] % umount /var/lib/waydroid/rootfs/vendor
(005909) [13:48:29] % umount /var/lib/waydroid/rootfs/vendor
(005909) [13:48:29] % umount /var/lib/waydroid/rootfs
(005909) [13:48:29] % umount /var/lib/waydroid/rootfs
```

`sudo waydroid shell -- logcat -d | tee logcat.txt`

```
[13:51:12] WayDroid container is STOPPED
```

`sudo dmesg | tee dmesg.txt`

Lots of

```
[  933.351364] binder_debug: 20 callbacks suppressed
[  933.351370] binder_linux: 5909:5909 cannot find target node
[  933.351378] binder_linux: 5909:5909 transaction call to 0:0 failed 12623/29189/-22, size 0-0 line 3044
[  933.351391] binder_linux: 6022:5913 cannot find target node
[  933.351399] binder_linux: 5913:6022 transaction call to 0:0 failed 12624/29189/-22, size 0-0 line 3044
[  933.996884] binder_linux: 3980:3980 cannot find target node
[  933.996894] binder_linux: 3980:3980 transaction call to 0:0 failed 12625/29189/-22, size 112-0 line 3044
[  934.352514] binder_linux: 5909:5909 cannot find target node
[  934.352525] binder_linux: 6022:5913 cannot find target node
[  934.352527] binder_linux: 5909:5909 transaction call to 0:0 failed 12626/29189/-22, size 0-0 line 3044
[  934.352533] binder_linux: 5913:6022 transaction call to 0:0 failed 12627/29189/-22, size 0-0 line 3044
```

**Comment by Djhg2000** [comment]
Looks related to #1165 . Raspberry Pi OS should have a Debian base so please paste the output of `grep BINDER /boot/config-$(uname -r)`.

**Comment by probonopd** [comment]
```
$ grep BINDER /boot/config-$(uname -r)
CONFIG_ANDROID_BINDER_IPC=y
CONFIG_ANDROID_BINDERFS=y
CONFIG_ANDROID_BINDER_DEVICES="binder,hwbinder,vndbinder"
# CONFIG_ANDROID_BINDER_IPC_SELFTEST is not set
```

**Comment by Djhg2000** [comment]
Alright, this one is a bit more tricky as I didn't notice the details of the errors in `dmesg` last time.

It looks like the binder interface is doing something weird but I have no clue how to debug this further without a local environment. I have a Pi 4 so I might be able to give it a try there when I find the time, probably not until next week at least.

If possible it would be interesting to see if it works on Ubuntu since I think that uses another build of the kernel, just to see if there's a problem with the kernel itself.

**Comment by aleasto** [comment]
I'm not familiar with LXDE, does it have a Wayland compositor now or are you nesting one?

You should be able to get the logcat if you start it quickly after `waydroid show-full-ui`, before the android container dies.

The most likely culprit is a missing kernel feature required by Android, and the logcat will show why.

**Comment by probonopd** [comment]
Raspberry Pi OS uses the Wayfire compositor.

**Comment by Botspot** [comment]
Many people are interested in this, especially with the release of the RPi5 which has plenty of power to run Waydroid well.
I would personally be thrilled to add this to our [Pi-Apps](https://github.com/Botspot/pi-apps) store.

@aleasto, On Raspberry Pi OS Bookworm which uses wayfire, I was able to prevent the "`Service manager /dev/binder has died`" error with these commands:
```
sudo apt install dkms
git clone https://github.com/choff/anbox-modules
cd anbox-modules/
./INSTALL.sh
```
However now `waydroid show-full-ui` only displays a black screen even after forcing software rendering as [explained here.](https://wiki.archlinux.org/title/Waydroid#Software_rendering)
@aleasto, any thoughts on this? Would it be better if I opened a new separate issue or continue discussion here? Would it help if I uploaded the logcat? Also, I would be glad to let you connect to my raspberry pi setup using a remote desktop software like WayVNC.

**Comment by choff** [comment]
Hello @Botspot ,

I think the logcat would definitely help to troubleshoot the issue.

While installing Waydroid on the Raspberry Pi 4, I saw thw following messages in the logcat output:
```
E libpsi  : No kernel psi monitor support (errno=2)
I lowmemorykiller: No kernel memory.pressure_level support (errno=2)
E lowmemorykiller: Kernel does not support memory pressure events or in-kernel low memory killer
I lowmemorykiller: exiting
```

I solved it by compiling a custom kernel for the Raspberry Pi with CONFIG_PSI enabled as described [here](https://www.raspberrypi.com/documentation/computers/linux_kernel.html). Afterwards, Waydroid worked fine, even though I could only test it with software rendering because there was a compile issue with the code for hardware rendering.

If enabling CONFIG_PSI helps, we could consider to file a request to the Raspberry Pi kernel team to build the kernel with this option per default.

I also don't see why you would need anbox-modules - at least if you are running a 64 bit kernel, binder should already be included (see [this issue](https://github.com/raspberrypi/linux/issues/4162)).

**Comment by Botspot** [comment]
> Hello @Botspot ,
> 
> I think the logcat would definitely help to troubleshoot the issue.
> 
> While installing Waydroid on the Raspberry Pi 4, I saw thw following messages in the logcat output:
> 
> ```
> E libpsi  : No kernel psi monitor support (errno=2)
> I lowmemorykiller: No kernel memory.pressure_level support (errno=2)
> E lowmemorykiller: Kernel does not support memory pressure events or in-kernel low memory killer
> I lowmemorykiller: exiting
> ```
> 
> I solved it by compiling a custom kernel for the Raspberry Pi with CONFIG_PSI enabled as described [here](https://www.raspberrypi.com/documentation/computers/linux_kernel.html). Afterwards, Waydroid worked fine, even though I could only test it with software rendering because there was a compile issue with the code for hardware rendering.
> 
> If enabling CONFIG_PSI helps, we could consider to file a request to the Raspberry Pi kernel team to build the kernel with this option per default.

I find this strange, as on PiOS Bookworm the kernel appears to already enable CONFIG_PSI=y by default.
I checked this with the following commands:
```
$ sudo modprobe configs
$ zgrep 'PSI' /proc/config.gz
CONFIG_PSI=y
CONFIG_PSI_DEFAULT_DISABLED=y
```
@choff, could you repeat your waydroid installation on the latest PiOS Bookworm 64-bit to see if the kernel still needs recompiling?

> I also don't see why you would need anbox-modules - at least if you are running a 64 bit kernel, binder should already be included (see [this issue](https://github.com/raspberrypi/linux/issues/4162)).

I understand that, but it seemed that without anbox-modules, the `Service manager /dev/binder has died` error would always occur. It's possible that something else I did is what actually solved that issue though.

**Comment by choff** [comment]
Hello @Botspot ,

yes, I want to also try out Waydroid on the Raspberry Pi 5, but that will take a longer time as I am not in possession of a Raspberry Pi 5 yet.. :-(

But good to know that CONFIG_PSI is already enabled - then I guess a custom kernel is likely not needed...


Best regards,

   Christian

**Comment by Botspot** [comment]
> Hello @Botspot ,
> 
> yes, I want to also try out Waydroid on the Raspberry Pi 5, but that will take a longer time as I am not in possession of a Raspberry Pi 5 yet.. :-(

I should have clarified. I am using RPi4 as well for these tests. If you got it working, I want to know how. (And why you had to enable psi when it seems it is already enabled for the rpi4 kernel)

**Comment by choff** [comment]
Hello @Botspot ,

sorry, I didn't read your message well.. :-(

Yes, that is actually the next step I had planned. I want to upgrade to Bookworm and also install Waydroid there. As of now, I am still on Debian Buster. I just checked again - on Buster, CONFIG_PSI is not enabled. But as you say, Bookworm now seems to enable this kernel option, so this should make it easier to run Waydroid easier. I will experiment with it around the Christmas time and will let you know the result.


Best regards,

   Christian

**Comment by probonopd** [comment]
I can do tests on a RPi 5 if you like.

**Comment by Botspot** [comment]
Here is the logcat:
In this test I was using a RPi4 on PiOS Bookworm 64-bit which uses Wayfire.
[logcat.txt](https://github.com/waydroid/waydroid/files/13688071/logcat.txt)

**Comment by Djhg2000** [comment]
@probonopd I already have a script for building Waydroid from source on a Debian-like distro, it's written primarily for Mobian but it might work on Raspberry Pi OS too if the dependencies are close enough. If the script is successful you will end up with Waydroid and the three dependencies missing in Debian installed as regular Debian packages.

As per usual I strongly recommend reading through the script before you run it. It requires `sudo` during some steps for installing dependencies but it's not using it for anything other than package management.

Just drop it in an empty directory and run (as a regular user, it will call sudo itself when needed): https://github.com/Djhg2000/Mobian-scripts/blob/master/Waydroid/rebuild_from_scratch.sh

You will have to go through the rest of the install steps after it's done, pretty much just `sudo waydroid init` and then either finding the desktop shortcut or running `waydroid show-full-ui`.

**Comment by probonopd** [comment]
Thanks a lot. Do you know why we need to compile this from source? I was hoping to be able to install some packages.

**Comment by aleasto** [comment]
I feel like the thread got sidetracked.
I'll quote myself:

> You should be able to get the logcat if you start it quickly after waydroid show-full-ui, before the android container dies.
> The most likely culprit is a missing kernel feature required by Android, and the logcat will show why.

**Comment by Botspot** [comment]
I am willing to allow a waydroid dev to connect to my Pi5 using [RustDesk,](https://rustdesk.com/) to try to pinpoint the cause of the crashes.
Just email me at `botspot@pi-apps.io` to exchange connection information. I am free for only the next few days.

**Comment by Djhg2000** [comment]
@probonopd I don't know if we _need_ to build it from source, I just wrote the script to quickly get the latest git version on Mobian.

**Comment by e100** [comment]
If you have a Raspberry PI 5 you need to:

Disable 16k pages by adding `kernel=kernel8.img` to `/boot/config.txt`:
`echo '# 4k pages
kernel=kernel8.img
'| sudo tee -a /boot/config.txt >/dev/null
`

Enable PSI by adding `psi=1` to `/boot/cmdline.txt`:
`sudo sed --follow-symlinks -i 's/quiet/psi=1 quiet/g' /boot/cmdline.txt
`

Then reboot

If you have a PI 4 or older, you likely just need to enable PSI,

**Comment by Botspot** [comment]
> Enable PSI by adding `psi=1` to `/boot/cmdline.txt`: `sudo sed --follow-symlinks -i 's/quiet/psi=1 quiet/g' /boot/cmdline.txt `

Is there any significance to adding psi=1 BEFORE `quiet`? I had it added at the end of the line.

**Comment by e100** [comment]
> Is there any significance to adding psi=1 BEFORE `quiet`? I had it added at the end of the line.

No, that is just how I choose to script the changes for my setup.
Also `/boot/cmdline.txt` is a symbolic link to the file `/boot/firmware/cmdline.txt` it is easy to accidentally change that symlink into a file such as using `sed -i` without the --follow-symlinks option.  If you are having issues check that the change is in the actual file `/boot/firmware/cmdline.txt`

If the psi=1 option is working, once you bootup you will find a folder `/proc/pressure`, if that folder is missing then psi=1 is missing.

**Comment by e100** [comment]
@Botspot 
> I should have clarified. I am using RPi4 as well for these tests. If you got it working, I want to know how. (And why you had to enable psi when it seems it is already enabled for the rpi4 kernel)

While the PI kernels do have PSI enabled, it is not turned on by default. Adding the kernel command line option `psi=1` will turn it on. Android will not run unless the kernel has PSI. The developers setup the kernel this way intentionally so PSI is not there when not needed and anyone who does need it can turn it on without having to recompile the kernel.

Please make sure you see my edited comment above, if you had copied/pasted the original command for disabling the 16k pages it did not work and you still have 16k pages enabled.
Android, currently, will only work on kernels with 4k pages and the PI 5 kernel defaults to using 16k pages making it impossible to run Android. The PI 4 and earlier kernels have 4k pages so this should only be an issue on the PI 5.

**Comment by Twix094** [comment]
I have a Pi 4B and a Pi 5 and I did some testing on both of them. In both cases I ended up with Waydroid launching but only showing a black screen and not being able to interact with Lineage OS at all.

Here's a screenshot of the result on the Pi 4B. Waydroid just hangs with the black screen until you right click on the taskbar and close the software or press ctrl+c in the terminal:
![20240103_14h03m48s_grim](https://github.com/waydroid/waydroid/assets/155034128/1dd9bca6-c5f9-4dc3-b6aa-775f8b5f90d0)

I'm assuming Waydroid is a x86_64 / amd64 program and therefore needs Box64 to run it on the Pi.

For the Pi 4B these are the steps I followed:
I did a fresh install of 2023-12-05-raspios-bookworm-arm64-full.img.xz downloaded from the Raspberry Pi website.
In the terminal I entered the following commands:
  `sudo sed --follow-symlinks -i 's/quiet/psi=1 quiet/g' /boot/cmdline.txt`
  `reboot`
I installed Pi-Apps with the following command:
  `wget -qO- https://raw.githubusercontent.com/Botspot/pi-apps/master/install | bash`
I opened Pi-Apps GUI and installed Box64, then added the repo for waydroid:
  `curl https://repo.waydro.id | sudo bash`
  `sudo apt install waydroid` (See Log1)
  `sudo box64 waydroid init`
 ` box64 waydroid session start`
This opened Waydroid to a black screen. I closed and stopped waydroid and tried again with with 'show-full-ui'.
  `waydroid session stop`
  `box64 waydroid show-full-ui` (See Log2)

As seen in Log1, during Waydroid install I got these messages:
```
Warning: unable to find a suitable fs in /proc/mounts, is it mounted?
Use --subdomainfs to override.
Failed to run 'apparmor_parser -rWT /etc/apparmor.d/lxc-containers'. You will pr
obably need to run it by hand at some point.
```

For the Pi 5, I followed these steps (and got the same result as the Pi 4B):
I did a fresh install of 2023-12-05-raspios-bookworm-arm64-full.img.xz downloaded from the Raspberry Pi website.
I ran the following command (copy and pasted from e100's github post):
```
  echo '# 4k pages
  kernel=kernel8.img
  '| sudo tee -a /boot/config.txt >/dev/null
```
Everything else after this is exactly the same as the Pi 4B.

Here are the Logs:
Log1: [Waydroid_Log1_Install.txt](https://github.com/waydroid/waydroid/files/13825376/Waydroid_Log1_Install.txt)
Log2: [Waydroid_Log2_ShowFullUI.txt](https://github.com/waydroid/waydroid/files/13825387/Waydroid_Log2_ShowFullUI.txt)

**Comment by Botspot** [comment]
@Twix094 Waydroid actually does have an arm64 release that runs without needing any emulation.
```
curl https://repo.waydro.id | sudo bash
sudo apt install waydroid -y
```

**Comment by Twix094** [comment]
I wasn't sure whether or not I needed Box64, it's good to know that I don't. 

I just repeated the whole process from my previous post on my Pi 5, except without installing Pi-Apps and Box64. I got exactly the same result, even when I take Box64 out of the equation.
Waydroid launches, but it's just a black screen. If I look in the program menu I can see all of the Lineage OS apps listed there. I can try to launch them, just to be greeted by the same black screen. It appears to be doing things in the background, it's just not displaying correctly.

**Comment by Botspot** [comment]
> I just repeated the whole process from my previous post on my Pi 5, except without installing Pi-Apps and Box64. I got exactly the same result, even when I take Box64 out of the equation. Waydroid launches, but it's just a black screen. If I look in the program menu I can see all of the Lineage OS apps listed there. I can try to launch them, just to be greeted by the same black screen. It appears to be doing things in the background, it's just not displaying correctly.

My findings are exactly the same here. Hopefully someone on this issue can help find out why. I am willing to help where possible.

**Comment by Mertica** [comment]
> > I just repeated the whole process from my previous post on my Pi 5, except without installing Pi-Apps and Box64. I got exactly the same result, even when I take Box64 out of the equation. Waydroid launches, but it's just a black screen. If I look in the program menu I can see all of the Lineage OS apps listed there. I can try to launch them, just to be greeted by the same black screen. It appears to be doing things in the background, it's just not displaying correctly.
> 
> My findings are exactly the same here. Hopefully someone on this issue can help find out why. I am willing to help where possible.

Same here on 4B after enabling PSI (which I wouldn't have even known about two days ago, when I tried researching this for the first time)

This "quick 5 minute setup" plan ended up with me researching until it was time to wake up to go to work, and still not having found a solution...

Stupid black screen...

Well, that's the life of Linux I guess...

**Comment by DienoX** [comment]
Does anyone have an idea where the problem could be?

**Comment by e100** [comment]
I had made many changes before even trying to install waydroid so maybe one of those changes is needed too.
I'm running my PI in X11 mode, not wayland and run waydroid inside of weston.  Maybe Wayfire is part of the issue too.
From the time I started installing waydroid the only additional changes I made to get it working were the psi=1 and the kernel with 4k pages.

Once I did get it running I noticed the errors described here, but it does run:
https://github.com/waydroid/waydroid/issues/1065

Applying the suggested fix did resolve those errors and seemed to result in android starting faster.

When I get home tonight I'll try to figure out a step by step process to get it working from a fresh image.

**Comment by e100** [comment]
Some of these steps are not needed like installing the editor I like to use but I included them to document every change I made.
TLDR; Wayfire is part of the problem

# Image
I used the Raspberry PI imager, selected PI 5 and Raspberry PI OS Bookworm 64bit

# Install preferred editor
```
sudo apt update
sudo apt install vim-nox
```

# Custom editor settings
```
echo 'unlet! skip_defaults_vim
source $VIMRUNTIME/defaults.vim
set mouse-=a
' > ~/.vimrc
```

# Fix usb power
My PI is powered by the header and I am booting from USB so I need this setting:

```
echo '# USB Power
usb_max_current_enable=1
'| sudo tee -a /boot/config.txt >/dev/null
```

# Change to kernel with 4K pages
```
echo '# 4k pages
kernel=kernel8.img
'| sudo tee -a /boot/config.txt >/dev/null
```

# Enable PSI
```
sudo sed --follow-symlinks -i 's/quiet/psi=1 quiet/g' /boot/cmdline.txt
```

# Install all the latest updates
```
sudo apt dist-upgrade -y
```

# Reboot

# Verify PSI is working and we have 4k pages:

This command should returtn '4096' to verify you are running a kernel with 4k pages::
```
getconf PAGESIZE
```

Verify the folder '/proc/pressure' exists to ensure PSI is working:
```
ls /proc/pressure
```

# Install Waydroid
```
sudo apt install curl ca-certificates -y
curl https://repo.waydro.id | sudo bash
sudo apt install -y waydroid
```

# Init Waydroid
```
sudo waydroid init
```

# First attempt to start waydroid
```
#waydroid --details-to-stdout show-full-ui
(012322) [17:39:28] Starting waydroid session
(012322) [17:39:28] Skipping clipboard manager service because of missing pyclip package
[gbinder] Service manager /dev/binder has appeared
(012322) [17:39:43] waydroidusermonitor: Received transaction: 1
(012322) [17:39:43] Android with user 0 is ready
```
I wait about 5 minutes, nothing happend and still blank screen

# Install weston
```
sudo apt install -y weston
```
First time I tried to run weston the window opened with corrupted graphics. Another sign that Wayfire is flawed<br>
So the next try I ran weston with this command:
```
weston -Bx11-backend.so
```

# Tried to start waydroid using weston:
In another terminal window I ran:
```
WAYLAND_DISPLAY=wayland-2 waydroid --details-to-stdout show-full-ui
```
Android started up and was working!

# Try something other than wayfire
I switched to gnome and it works without having to run weston
```
sudo apt install -y gnome-session
```

Run this command and select gnome-session:
```
sudo update-alternatives --config x-session-manager
```
reboot

**Comment by Mertica** [comment]
> Some of these steps are not needed like installing the editor I like to use but I included them to document every change I made. TLDR; Wayfire is part of the problem

I didn't try the GNOME steps, but everything above it seemed to make it work, I think mainly the Weston part

Thank you!

**Comment by Botspot** [comment]
I don't have the time to set up a waydroid install right now, but if someone else would like to try this fix, it might get waydroid working under wayfire.
`waydroid prop set persist.waydroid.multi_windows true`
See: https://github.com/WayfireWM/wayfire/issues/1615#issuecomment-1384790916

**Comment by Botspot** [comment]
Update: that does indeed fix the black screen issue. First run `waydroid show-full-ui` as normal, then while that is running, run `waydroid prop set persist.waydroid.multi_windows true`. Now kill the `waydroid show-full-ui` process, and restart it. See that it now works.

The wayfire dev just responded with a patch for wayfire which should make the workaround unnecessary. I don't have the time to figure that out, but if someone else here would like to recompile [the Pi OS fork of wayfire](https://github.com/raspberrypi-ui/wayfire), it should be pretty easy to apply this patch:
```
diff --git a/src/view/surface.cpp b/src/view/surface.cpp
index 390ec0de..2b2f087d 100644
--- a/src/view/surface.cpp
+++ b/src/view/surface.cpp
@@ -45,6 +45,15 @@ wf::wlr_surface_controller_t::wlr_surface_controller_t(wlr_surface *surface,
         // Allocate memory, it will be auto-freed when the wlr objects are destroyed
         auto sub_controller = new wlr_subsurface_controller_t(sub);
         create_controller(sub->surface, sub_controller->get_subsurface_root());
+        wlr_subsurface *s;
+        wl_list_for_each(s, &surface->current.subsurfaces_below, current.link)
+        {
+            if (sub == s)
+            {
+                wf::scene::add_back(this->root, sub_controller->get_subsurface_root());
+                return;
+            }
+        }
         wf::scene::add_front(this->root, sub_controller->get_subsurface_root());
     });
 ```
Then of course report back your findings and leave a comment on [the original wayfire issue](https://github.com/WayfireWM/wayfire/issues/1615).

**Comment by 1ubuntuuser** [comment]
@Botspot thanks for the heads up. Does the system support HW encoding/decoding of video?

**Comment by DienoX** [comment]
Raspbery pi 5 Does it only have hardware support for H265?

**Comment by aleasto** [comment]
Closing this as it was correctly triaged. You need:
* A kernel with 4K page sizes until Android 15
* A kernel with PSI enabled
* Wayfire updated with the patch to correctly implement the wl_subsurface protocol

**Comment by probonopd** [comment]
So does this essentially mean "no Waydroid for Raspberry Pi OS anytime soon"?

**Comment by aleasto** [comment]
Kernel configuration is outside of our control.  At the moment 4K pages and PSI are a hard requirement.
Wayfire just needs a tagged release.

**Comment by Botspot** [comment]
> So does this essentially mean "no Waydroid for Raspberry Pi OS anytime soon"?

Raspberry Pi Wayfire needs to be updated to pull in the patch to fix the black window issue. I have just requested they do it in this issue here: https://github.com/raspberrypi-ui/wayfire/issues/7
@probonopd, Waydroid actually can work on default PiOS; you just need to do [this workaround](https://github.com/waydroid/waydroid/issues/1171#issuecomment-1894993339) which I described earlier. That workaround will not be necessary once PiOS wayfire pulls in the patch.

**Comment by Botspot** [comment]
@probonopd, a RPi dev responded to my issue and said that they eventually plan to update wayfire, but it won't be easy or soon. https://github.com/raspberrypi-ui/wayfire/issues/7#issuecomment-1951179967

**Comment by olevenets2** [comment]
> Update: turns out I never uninstalled waydroid and it was ready to go. The workaround I mentioned above does indeed fix the black screen issue! First run `waydroid show-full-ui` as normal, then while that is running, run `waydroid prop set persist.waydroid.multi_windows true`. Now kill the `waydroid show-full-ui` process, and restart it. See that it now works.
> 
> The wayfire dev just responded with a patch for wayfire which should make the workaround unnecessary. I don't have the time to figure that out, but if someone else here would like to recompile [the Pi OS fork of wayfire](https://github.com/raspberrypi-ui/wayfire), it should be pretty easy to apply this patch:
> 
> ```
> diff --git a/src/view/surface.cpp b/src/view/surface.cpp
> index 390ec0de..2b2f087d 100644
> --- a/src/view/surface.cpp
> +++ b/src/view/surface.cpp
> @@ -45,6 +45,15 @@ wf::wlr_surface_controller_t::wlr_surface_controller_t(wlr_surface *surface,
>          // Allocate memory, it will be auto-freed when the wlr objects are destroyed
>          auto sub_controller = new wlr_subsurface_controller_t(sub);
>          create_controller(sub->surface, sub_controller->get_subsurface_root());
> +        wlr_subsurface *s;
> +        wl_list_for_each(s, &surface->current.subsurfaces_below, current.link)
> +        {
> +            if (sub == s)
> +            {
> +                wf::scene::add_back(this->root, sub_controller->get_subsurface_root());
> +                return;
> +            }
> +        }
>          wf::scene::add_front(this->root, sub_controller->get_subsurface_root());
>      });
> ```
> 
> Then of course report back your findings here and leave a comment on [the original wayfire issue](https://github.com/WayfireWM/wayfire/issues/1615).

Hello, is there a patch for wayfire 0.7.5? I could not apply it to the version of the program from Raspberry Pi OS
